lib/Dialect/Ibis/IbisOps.cpp

```diff
@@ -270,24 +286,114 @@ LogicalResult ThisOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
 // ContainerInstanceOp
 //===----------------------------------------------------------------------===//
 
-ContainerOp
-ContainerInstanceOp::getContainer(SymbolTableCollection *symbolTable) {
+ContainerOp ContainerInstanceOp::getContainer(SymbolTable *symbolTable) {
   auto mod = getOperation()->getParentOfType<mlir::ModuleOp>();
   if (symbolTable)
-    return symbolTable->lookupSymbolIn<ContainerOp>(mod, getTargetNameAttr());
+    return dyn_cast<ContainerOp>(
+        symbolTable->lookupSymbolIn(mod, getTargetNameAttr()));
 
   return mod.lookupSymbol<ContainerOp>(getTargetNameAttr());
 }
 
 LogicalResult
 ContainerInstanceOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-  auto targetContainer = getContainer(&symbolTable);
+  auto targetContainer = getContainer(&symbolTable.getSymbolTable(
+      getOperation()->getParentOfType<mlir::ModuleOp>()));
   if (!targetContainer)
     return emitOpError() << "'" << getTargetName() << "' does not exist";
 
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// PathOp
+//===----------------------------------------------------------------------===//
+
+/// Infer the return types of this operation.
+LogicalResult PathOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> loc, ValueRange operands,
+    DictionaryAttr attrs, mlir::OpaqueProperties properties,
+    mlir::RegionRange regions, SmallVectorImpl<Type> &results) {
+  auto path = attrs.get("path").cast<ArrayAttr>();
+  if (path.empty())
+    return failure();
+
+  auto lastStep = path.getValue().back().cast<PathStepAttr>();
+  results.push_back(lastStep.getType());
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// PathOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult PathStepAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                                   PathDirection direction, mlir::Type type,
+                                   mlir::FlatSymbolRefAttr instance) {
+  // 'parent' should never have an instance name specified.
+  if (direction == PathDirection::Parent && instance)
+    return emitError() << "ibis.step 'parent' may not specify an instance name";
+
+  if (direction == PathDirection::Child && !instance)
+    return emitError() << "ibis.step 'child' must specify an instance name";
+
+  // Only allow scoperefs
+  auto scoperefType = type.dyn_cast<ScopeRefType>();
+  if (!scoperefType)
+    return emitError() << "ibis.step type must be an !ibis.scoperef type";
+
+  return success();
+}
+
+LogicalResult PathOp::verify() {
+  auto pathRange = getPathAsRange();
+  if (pathRange.empty())
+    return emitOpError() << "ibis.path must have at least one step";
```

Comment:
I think this verification is fine, but a path with no steps would just refer to `this`, correct?